editor/src/core/tailwind/tailwind.ts

```diff
@@ -0,0 +1,243 @@
+    if (isRight(maybeConfig)) {
+      return maybeConfig.value
+    } else {
+      return {}
+    }
+  }, [tailwindConfigFile, requireFn])
+}
+
+interface TwindInstance {
+  element: HTMLStyleElement
+  observer: TwindObserver
+}
+
+let twindInstance: TwindInstance | null = null
+
+function clearTwind() {
+  if (twindInstance != null) {
+    twindInstance.observer.disconnect()
+    twindInstance.element.parentNode?.removeChild(twindInstance.element)
+  }
+}
+
+const adjustRuleScope = memoize(
+  (rule: string, prefixSelector: string | null): string => {
+    if (prefixSelector == null) {
+      return rule
+    } else {
+      const splitOnBrace = rule.split('{')
+      const splitSelectors = splitOnBrace[0].split(',')
+      const scopedSelectors = splitSelectors.map((s) => {
+        const lowerCaseSelector = s.toLowerCase().trim()
+
+        if (
+          lowerCaseSelector === ':root' ||
+          lowerCaseSelector === 'html' ||
+          lowerCaseSelector === 'head'
+        ) {
+          return prefixSelector
+        } else if (lowerCaseSelector === 'body') {
+          return `${prefixSelector} > *`
+        } else {
+          return `${prefixSelector} ${s}`
+        }
+      })
+      const joinedSelectors = scopedSelectors.join(',')
+      const afterBrace = splitOnBrace.slice(1)
+      const finalRule = [joinedSelectors, ...afterBrace].join('{')
+      return finalRule
+    }
+  },
+  { maxSize: 100, equals: (a, b) => a === b },
+)
+
+function updateTwind(config: Configuration, prefixSelector: string | null) {
+  const element = document.head.appendChild(document.createElement('style'))
+  element.appendChild(document.createTextNode('')) // Avoid Edge bug where empty style elements doesn't create sheets
+
+  const sheet = cssomSheet({ target: element.sheet ?? undefined })
+  const customSheet: Sheet = {
+    target: sheet.target,
+    insert: (rule, index) => {
+      const scopedRule = adjustRuleScope(rule, prefixSelector)
+      sheet.insert(scopedRule, index)
+    },
+  }
+
+  clearTwind()
+
+  const observer = observe(
+    document.documentElement,
+    create({ ...config, sheet: customSheet, mode: silent }),
+  )
+
+  twindInstance = {
+    element: element,
+    observer: observer,
+  }
+}
+
+export function useTwind(
+  projectContents: ProjectContentTreeRoot,
+  requireFn: RequireFn,
+  prefixSelector: string | null = null,
+) {
+  const hasDependencies = useHasRequiredDependenciesForTailwind(projectContents)
+  const hasPostCSSPlugin = usePostCSSIncludesTailwindPlugin(projectContents, requireFn)
+  const shouldUseTwind = hasDependencies && hasPostCSSPlugin
+  const tailwindConfig = useGetTailwindConfig(projectContents, requireFn)
+  React.useMemo(() => {
+    if (shouldUseTwind) {
+      updateTwind(tailwindConfig, prefixSelector)
+    } else {
+      clearTwind()
+    }
+  }, [prefixSelector, shouldUseTwind, tailwindConfig])
+}
+
+export function injectTwind(
```

Comment:
This is just a non-hooks version of the above for use in the preview because the preview render function (which needs to call this) isn't a React component. It's a bit rubbish having to effectively duplicate the logic like this :/